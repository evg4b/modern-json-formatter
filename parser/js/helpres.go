package js

func errorResponse(error string) map[string]any {
	return map[string]any{
		"type":  "error",
		"error": error,
	}
}

func successResponse(value map[string]any) map[string]any {
	return map[string]any{
		"type":  "response",
		"value": value,
	}
}
