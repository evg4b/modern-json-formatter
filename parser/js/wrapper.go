package js

import (
	"github.com/evg4b/modern-json-formatter/parser/decode"
	"syscall/js"
)

func ParseWrapper() js.Func {
	return js.FuncOf(func(_ js.Value, args []js.Value) any {
		if len(args) != 1 {
			return map[string]interface{}{
				"type":  "error",
				"error": "Invalid arguments passed",
			}
		}

		jsonTree, err := decode.Decode(args[0].String())
		if err != nil {
			return map[string]any{
				"type":  "error",
				"error": err.Error(),
			}
		}

		return jsonTree
	})
}
