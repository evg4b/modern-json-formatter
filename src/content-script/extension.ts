import { jq, tokenize, type TokenizerResponse } from '@core/background';
import { registerStyles } from '@core/ui/helpers';
import { isNotNull } from 'typed-assert';
import { buildDom } from './dom';
import { buildErrorNode } from './dom/build-error-node';
import { findNodeWithCode } from './json-detector';
import styles from './styles.module.scss';
import { buildContainers } from './ui/containers';
import { ToolboxElement } from './ui/toolbox';

export const runExtension = async () => {
  const preNode = await findNodeWithCode();
  if (!preNode) {
    return;
  }

  const shadowRoot = document.body.attachShadow({ mode: 'open' });
  registerStyles(shadowRoot, styles);

  const content = preNode.textContent;
  isNotNull(content, 'No data found');

  const { rootContainer, rawContainer, formatContainer, queryContainer } = buildContainers(shadowRoot);
  rawContainer.appendChild(preNode);

  const wrapper = async <T>(promise: Promise<T>): Promise<T> => {
    rootContainer.classList.add('loading');
    try {
      return await promise;
    } finally {
      rootContainer.classList.remove('loading');
    }
  };

  const toolbox = new ToolboxElement();
  shadowRoot.appendChild(toolbox);

  toolbox.onQueryChanged(async query => {
    await wrapper(
      (async (query: string) => {
        const info = await jq(preNode.innerText, query);
        if (info.type === 'error' && info.scope === 'jq') {
          toolbox.setErrorMessage(info.error);
          return;
        }

        queryContainer.innerHTML = '';
        queryContainer.appendChild(prepareResponse(info));
      })(query),
    );
  });

  toolbox.onTabChanged(tab => {
    switch (tab) {
      case 'query':
        rootContainer.classList.remove('raw', 'formatted');
        rootContainer.classList.add('query');
        return;
      case 'raw':
        rootContainer.classList.remove('formatted', 'query');
        rootContainer.classList.add('raw');
        return;
      case 'formatted':
        rootContainer.classList.remove('raw', 'query');
        rootContainer.classList.add('formatted');
        return;
    }
  });

  await wrapper(tokenize(preNode.innerText).then(response => formatContainer.appendChild(prepareResponse(response))));
};

const prepareResponse = (response: TokenizerResponse): HTMLElement => {
  return response.type === 'error' ? buildErrorNode('Invalid JSON file.', response.error) : buildDom(response);
};
