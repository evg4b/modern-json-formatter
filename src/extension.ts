import { ErrorNode, TokenizerResponse } from '@packages/tokenizer';
import { isNotNull } from 'typed-assert';
import { buildDom } from './dom';
import { buildErrorNode } from './dom/build-error-node';
import { detectJson, getJsonSelector } from './json-detector/detect';
import styles from './styles.scss';
import { buildButtons } from './ui/buttons';
import { buildContainers } from './ui/containers';

export const runExtension = async () => {
  if (!await detectJson()) {
    return;
  }

  const { tokenize } = await import('@packages/tokenizer');

  const shadow = document.body.attachShadow({ mode: 'open' });
  const styleNode = document.createElement('style');
  styleNode.textContent = styles;
  styleNode.setAttribute('type', 'text/css');
  styleNode.setAttribute('rel', 'stylesheet');
  styleNode.setAttribute('role', 'presentation');

  shadow.appendChild(styleNode);

  const data = document.querySelector<HTMLPreElement>(getJsonSelector());
  isNotNull(data, 'No data found');

  const { rootContainer, rawContainer, formatContainer, queryContainer } = buildContainers(shadow);
  rawContainer.appendChild(data);

  const { rawButton, queryButton, formatButton, queryInput } = buildButtons(shadow);

  const response = await tokenize(data.innerText);

  queryInput.style.display = 'none';
  rawButton.addEventListener('click', () => {
    rootContainer.classList.remove('formatted', 'query');
    rootContainer.classList.add('raw');
    queryInput.style.display = 'none';
  });

  formatButton.addEventListener('click', () => {
    rootContainer.classList.remove('raw', 'query');
    rootContainer.classList.add('formatted');
    queryInput.style.display = 'none';
  });

  queryButton.addEventListener('click', () => {
    rootContainer.classList.remove('raw', 'formatted');
    rootContainer.classList.add('query');
    queryInput.style.display = 'inline-block';
  });

  // eslint-disable-next-line @typescript-eslint/no-misused-promises
  queryInput.addEventListener('keydown', async event => {
    if (event.key === 'Enter') {
      const { jq } = await import('@packages/jq');

      const info = await jq(data.innerText, queryInput.value);
      queryContainer.innerHTML = '';
      queryContainer.appendChild(
        prepareResponse(info),
      );
    }
  });

  formatContainer.appendChild(
    prepareResponse(response),
  );
};

const prepareResponse = (response: TokenizerResponse): HTMLElement => {
  return response.type === 'error'
    ? buildErrorNode(extractHeader(response), extractLines(response))
    : buildDom(response);
};

const extractLines = (response: ErrorNode) => response.error;

const extractHeader = (response: ErrorNode): string => response.scope === 'tokenizer'
  ? 'Invalid JSON file.'
  : 'Invalid JQ Query.';


