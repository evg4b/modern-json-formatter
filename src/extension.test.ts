import { readFile } from 'fs/promises';
import { resolve } from 'node:path';
import { runExtension } from './extension';
import '../parser/wasm_exec.js';

jest.mock('./parser', () => ({
  initParser: jest.fn().mockReturnValueOnce(Promise.resolve()),
}));

function buildChromeJsonPage(json: string) {
  document.body.innerHTML = ``;
  const pre = document.createElement('pre');
  pre.innerText = json;
  document.body.appendChild(pre);
  const container = document.createElement('div');
  container.classList.add('json-formatter-container');
  document.body.appendChild(container);
}

describe('extension', () => {
  const originalFetch = global.fetch;
  let go: Go;

  beforeEach(() => {
    Object.assign(global, {
      chrome: {
        runtime: {
          getURL: jest.fn(() => 'parser.wasm'),
        },
      },
    });
  });

  beforeAll(async () => {
    const wasmBuffer = await readFile(resolve(__dirname, '../parser/parser.wasm'));
    go = new Go();
    const wasm = await WebAssembly.instantiate(wasmBuffer, go.importObject);
    go.run(wasm.instance);
  });

  afterAll(() => {
    global.fetch = originalFetch;
  });

  describe('for valid json', () => {
    beforeEach(() => {
      buildChromeJsonPage(`{"userId": 1,"id": 1,"title": "delectus aut autem","completed": false}`);
    });

    test('should render extension', async () => {
      await runExtension();

      expect(document.body.shadowRoot?.innerHTML).toMatchSnapshot();
    });
  });
});
