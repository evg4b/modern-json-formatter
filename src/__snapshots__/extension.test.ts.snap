// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extension should run 1`] = `"<style></style><div class="root-container"><div class="formatted-json-container"><span class="root"><span class="toggle"></span><span class="object"><span class="bracket">{</span><span class="inner"><span class="property"><span class="key">"title"</span><span class="colon">:</span><span class="string">"delectus aut autem"</span><span class="comma">,</span></span><span class="property"><span class="key">"completed"</span><span class="colon">:</span><span class="boolean">false</span><span class="comma">,</span></span><span class="property"><span class="key">"userId"</span><span class="colon">:</span><span class="number">1</span><span class="comma">,</span></span><span class="property"><span class="key">"id"</span><span class="colon">:</span><span class="number">1</span></span></span><span class="ellipsis"></span><span class="bracket">}</span></span></span></div><div class="raw-json-container"><pre></pre></div></div><div class="button-container"><button class="active">Formatted</button><button>Raw</button></div>"`;
